File: scripts/pseudo-words.py
import json
from collections import Counter
import random

# Read the trigrams file
with open("assets/trigrams.json", "r", encoding="utf-8") as substrings:
    trigram_data = json.load(substrings)

global bigram_data
# Read the bigrams file
with open("assets/bigrams.json", "r", encoding="utf-8") as bigrams:
    bigram_data = json.load(bigrams)
# Pick a random bigram (with weight)
total_bi_weight = sum(bigram_data.values())
bi_weights = [count/total_bi_weight for count in bigram_data.values()]
# Pick the bigram and use it as the initial string
string = random.choices(list(bigram_data.keys()), weights = bi_weights, k = 1)[0]

counter = 0 # Counter for the amount of times to iterate
def generate_word(iterations):
    global string
    global counter

    total_matches = 0
    matches = {}
    for trigram in trigram_data: # Get all the trigrams with the first two letters that match the last two letters of the string
        if trigram.startswith(string[-2:]):
            matches.update({trigram: trigram_data[trigram]}) # Add to a dictionary
            total_matches += int(trigram_data[trigram]) # Add values to one variable for weighted probability

    if matches:
        weights = [matches[trigram] / total_matches for trigram in matches] # Weights for the random choice
        selected_trigram = random.choices(list(matches.keys()), weights = weights, k = 1)[0] # Select trigram randomly

        string = string[:-2] # Remove last two letters of the string before adding the trigram
        string += selected_trigram # Add the trigram to the string
        if counter <= iterations: # Run multiple times
            counter += 1
            generate_word(iterations - 1)

# Create multiple words and store them in a dictionary
word_dict = {}
for i in range(10000):  # Generate 10000 words
    string = random.choices(list(bigram_data.keys()), weights = bi_weights, k = 1)[0]  # Reset string
    counter = 0  # Reset counter
    generate_word(random.randint(2, 7))  # Generate new word
    word_dict[str(i)] = string  # Add to dictionary with index as key

# Write to file
with open("assets/pseudo-words.json", "w") as substrings:
    json.dump(word_dict, substrings, indent = 4)

File: lib/pseudo-words.js
let trigrams
let bigrams
let endings
let word
let string = ""

export class pseudoWords {
    static iterations = 5
    static capitalise = true    
    static punctuation = false
    static amount = 10

}

async function getTextData() {
    try {
        // Wait for both fetches to complete
        const [trigramsResponse, bigramsResponse, endingsResponse] = await Promise.all([
            fetch("assets/trigrams.json"),
            fetch("assets/bigrams.json"),
            fetch("assets/endings.json")
        ])

        // Convert both responses to JSON
        trigrams = await trigramsResponse.json()
        bigrams = await bigramsResponse.json()
        endings = await endingsResponse.json()

        return {trigrams, bigrams, endings}

    }   catch (error) {
        console.error('Error fetching data:', error)
    }
}

// Correct way to use the async function
export async function createString() {
    const textData = await getTextData()

    // Get the total weights of the bigrams and trigrams
    const totalBiWeights = Object.values(textData.bigrams).reduce((sum, current) => sum + current, 0)
    const totalTriWeights = Object.values(textData.trigrams).reduce((sum, current) => sum + current, 0)
    const totalEndWeights = Object.values(textData.endings).reduce((sum, current) => sum + current, 0)

    for (let i = 0; i < pseudoWords.amount; i++) {
        // Pick a random initial bigram
        let weightSum = 0
        let initialBigram
        let random_ = Math.random() * totalBiWeights
        for (const [bigram, weight] of Object.entries(textData.bigrams)) {
            // https://stackoverflow.com/questions/43566019/how-to-choose-a-weighted-random-array-element-in-javascript
            // Add accumalating weight to current bigram
            weightSum += weight
            if (random_ <= weightSum) {
                // If the number chosen is less than or equal the sum of the weights before current bigram, pick said bigram
                initialBigram = bigram
                break
            }
        }

        // Create first two letters of the word
        pseudoWords.capitalise ? word = initialBigram.slice(0,1).toUpperCase() + initialBigram.slice(1) : word = initialBigram

        // Iterate "i" times on word, getting the last two letters of it every time and matching it with a trigram, then adding the last letter of the trigram to the word
        for (let i = 0; i < pseudoWords.iterations - 3; i++) {
            // Reset matching trigrams object every iteration
            let matchingTrigrams = {}
            let matchingTriWeights = 0
            Object.entries(textData.trigrams).forEach(([trigram, weight]) => {
                // Check for each trigram that starts with the same two letters as the end of the word and create an object for it
                if (trigram.startsWith(word.toLowerCase().slice(-2))) {
                    matchingTrigrams[trigram] = weight
                    matchingTriWeights += weight
                }
            })
            
            let weightSum = 0
            // Pick a random trigram from matching trigrams (weighted choice)
            let random = Math.random() * matchingTriWeights
            for (const [matchingTrigram, weight] of Object.entries(matchingTrigrams)) {
                // https://stackoverflow.com/questions/43566019/how-to-choose-a-weighted-random-array-element-in-javascript
                // Add accumalating weight to current bigram
                weightSum += weight
                if (random <= weightSum) {
                    // If the number chosen is less than or equal the sum of the weights before current trigram, pick said trigram
                    // Add trigram to the end of the word
                    word += matchingTrigram.slice(-1)
                    break
                }
            }
        }

        // Use the word endings for the last three letters of the word
            // Reset matching trigrams object every iteration
            let matchingEndings = {}
            let matchingEndWeights = 0
            Object.entries(textData.endings).forEach(([ending, weight]) => {
                // Check for each trigram that starts with the same two letters as the end of the word and create an object for it
                if (ending.startsWith(word.toLowerCase().slice(-2))) {
                    matchingEndings[ending] = weight
                    matchingEndWeights += weight
                }
            })

        // Pick a random trigram from matching trigrams (weighted choice)
        let random = Math.random() * matchingEndWeights
        for (const [matchingEnding, weight] of Object.entries(matchingEndings)) {
            // https://stackoverflow.com/questions/43566019/how-to-choose-a-weighted-random-array-element-in-javascript
            // Add accumalating weight to current bigram
            weightSum += weight
            if (random <= weightSum) {
                // If the number chosen is less than or equal the sum of the weights before current trigram, pick said trigram
                // Add trigram to the end of the word
                word += matchingEnding.slice(-3)
                break
            }
        }

        if (pseudoWords.punctuation === true) {
            const punctuation = [".", ",", "!", "?", ";"]
            word += punctuation[Math.floor(Math.random() * punctuation.length)]
        }

        word += " "
        string += word

    }
    return string
    
}